File: web/src/boss-expression.ts
export type BossExpression = "normal" | "satisfied" | "displeased" | "encouraging";

/**
 * resolveBossExpression の入力。GET /api/dashboard（#58）のレスポンスから
 * 必要なフィールドだけを抜き出したフラットな独自型（API 型に依存させない）。
 */
export interface BossExpressionContext {
  progressRatio: number;
  morningSessionHeld: boolean;
  eveningSessionHeld: boolean;
  todayMaxEscalationLevel: number;
}

const ESCALATION_DISPLEASED_THRESHOLD = 2;
const SATISFIED_RATIO_THRESHOLD = 0.8;
const EVENING_DISPLEASED_RATIO_THRESHOLD = 0.5;
const ENCOURAGING_RATIO_THRESHOLD = 0.3;

/**
 * 文脈からボスの表情を決める純粋関数。優先順位（Issue #59 の明示的な仮定）:
 * 1. エスカレーションレベル >= 2 → displeased
 * 2. 夕会済みかつ達成率 >= 0.8 → satisfied
 * 3. 夕会済みかつ達成率 < 0.5 → displeased
 * 4. 達成率 >= 0.8 → satisfied
 * 5. 朝会済みかつ達成率 < 0.3 → encouraging
 * 6. それ以外 → normal
 */
export function resolveBossExpression(
  context: BossExpressionContext,
): BossExpression {
  const {
    progressRatio,
    morningSessionHeld,
    eveningSessionHeld,
    todayMaxEscalationLevel,
  } = context;

  if (todayMaxEscalationLevel >= ESCALATION_DISPLEASED_THRESHOLD) {
    return "displeased";
  }
  // ルール4（達成率 >= 0.8 → satisfied）に完全に包含されるため実行上は冗長だが、
  // Issue #59 が明示する優先順位リストとの1対1対応を保ちトレーサビリティを優先する。
  if (eveningSessionHeld && progressRatio >= SATISFIED_RATIO_THRESHOLD) {
    return "satisfied";
  }
  if (eveningSessionHeld && progressRatio < EVENING_DISPLEASED_RATIO_THRESHOLD) {
    return "displeased";
  }
  if (progressRatio >= SATISFIED_RATIO_THRESHOLD) {
    return "satisfied";
  }
  if (morningSessionHeld && progressRatio < ENCOURAGING_RATIO_THRESHOLD) {
    return "encouraging";
  }
  return "normal";
}

File: spikes/flutter/app/lib/dashboard/boss_expression.dart
// web/src/boss-expression.ts / dashboard-progress-level.ts / evening-evaluation.ts の移植。
enum BossExpression { normal, satisfied, displeased, encouraging }

class BossExpressionContext {
  const BossExpressionContext({
    required this.progressRatio,
    required this.morningSessionHeld,
    required this.eveningSessionHeld,
    required this.todayMaxEscalationLevel,
  });
  final double progressRatio;
  final bool morningSessionHeld;
  final bool eveningSessionHeld;
  final int todayMaxEscalationLevel;
}

const escalationDispleasedThreshold = 2;
const satisfiedRatioThreshold = 0.8;
const eveningDispleasedRatioThreshold = 0.5;
const encouragingRatioThreshold = 0.3;

BossExpression resolveBossExpression(BossExpressionContext c) {
  if (c.todayMaxEscalationLevel >= escalationDispleasedThreshold) return BossExpression.displeased;
  if (c.eveningSessionHeld && c.progressRatio >= satisfiedRatioThreshold) return BossExpression.satisfied;
  if (c.eveningSessionHeld && c.progressRatio < eveningDispleasedRatioThreshold) return BossExpression.displeased;
  if (c.progressRatio >= satisfiedRatioThreshold) return BossExpression.satisfied;
  if (c.morningSessionHeld && c.progressRatio < encouragingRatioThreshold) return BossExpression.encouraging;
  return BossExpression.normal;
}

enum ProgressLevel { low, mid, high }

ProgressLevel resolveProgressLevel(double ratio) {
  if (ratio < encouragingRatioThreshold) return ProgressLevel.low;
  if (ratio >= satisfiedRatioThreshold) return ProgressLevel.high;
  return ProgressLevel.mid;
}

enum EveningEvaluationTone { praise, scold, neutral }

EveningEvaluationTone resolveEveningEvaluationTone(double ratio) {
  if (ratio >= satisfiedRatioThreshold) return EveningEvaluationTone.praise;
  if (ratio < eveningDispleasedRatioThreshold) return EveningEvaluationTone.scold;
  return EveningEvaluationTone.neutral;
}
